nodejs/week3/src/backend/api/meals.js

```diff
@@ -0,0 +1,148 @@
+    throw error;
+  }
+});
+
+router.get("/titles", async (request, response) => {
+  try {
+    // knex syntax for selecting table records.
+    const titles = await knex("meals").select("title");
+    response.json(titles);
+  } catch (error) {
+    throw error;
+  }
+});
+
+//GET api/meals/ query parameters
+
+router.get("/", async function (req, res) {
+  const {
+    maxPrice,
+    title,
+    availableReservations,
+    createdAfter,
+    limit,
+  } = req.query;
+  const mealTitles = await knex.select("*").table("meals");
+
+  if (Object.keys(req.query).length === 0) {
+    res.send(mealTitles);
+  }
+
+  if (maxPrice) {
+    try {
+      const maxPriceMeals = await knex(meals)
+        .select("*")
+        .where("price", "<", maxPrice);
+      console.log(maxPriceMeals);
+      res.send(maxPriceMeals);
+    } catch (error) {
+      error;
+    }
+  } else if (title) {
+    try {
+      const titles = await knex(meals)
+        .select("*")
+        .where("title", "like", `%${title}%`);
+      console.log(titles);
+      res.send(titles);
+    } catch (error) {
+      error;
+    }
+  } else if (availableReservations === "true") {
+    try {
+      const availableReservationsMeal = await knex
+        .select("*")
+        .sum({
+          sumTotal: "reservations.number_of_guests",
+        })
+        .table("meals")
+        .join("reservations", {
+          "meals.id": "reservations.meal_id",
+        })
+        .groupBy("meals.title")
+        .having("sumTotal", ">", "Max_reservations");
+      console.log(availableReservationsMeal);
+      res.send(availableReservationsMeal);
+    } catch (error) {
+      error;
+    }
+  } else if (createdAfter) {
+    try {
+      const datecreated = await knex
+        .select("*")
+        .table("meals")
+        .where("created_date", "<", createdAfter);
+      console.log(datecreated);
+      res.send(datecreated);
+    } catch (error) {
+      error;
+    }
+  } else if (limit) {
+    try {
+      const limitMeals = await knex.select("*").table("meals").limit(limit);
+      console.log(limitMeals);
+      res.send(limitMeals);
+    } catch (error) {
+      error;
+    }
+  }
+});
+
+//Adds a new meal
+router.post("/", async (req, res) => {
+  const meal = {
+    id: req.body.id,
+    title: req.body.title,
+    description: req.body.description,
+    location: req.body.location,
+    when: req.body.when,
+    Max_reservation: req.body.Max_reservation,
+    price: req.body.price,
+    created_date: req.body.created_date,
+  };
+  await knex("meals").insert(req.body);
+});
+
+//Get meals with id
+router.get("/:id", async (req, res) => {
+  try {
+    const { id } = req.params;
+    const mealWithIDb = await knex("meals").where({ id });
+    res.json(mealWithIDb);
+  } catch (error) {
+    throw error;
+  }
+});
+
+//Updates the meal by id
+router.put("/:id", async (req, res) => {
+  try {
+    const updatedMeal = await knex("meals").where(req.params).update(req.body);
```

Comment:
when you look at this code, it is kind of hard to understand at a glance what is the where clause in this query. It requires a cognitive leap to think of the parameters of the request, go and take a look and identify that it is an `:id`. I would create a variable to be more explicit:

`const {id} = req.params`

then

`await knext('meals').where({id})...`

That way, when a person reads this code it is very clearly to understand that you're querying by id. You don't need to jump around the code trying to find what exactly the `req.params` are.

When writing code in the future think about how you can make the life of the code reader easier. Because code is read a lot more times that it is written.